samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java

```diff
@@ -16,85 +16,378 @@
+   * {@link org.apache.samza.operators.triggers.Triggers} APIs.
+   *
+   * For example, to window a stream into count based windows that trigger every 50 messages or every 10 minutes.
+   * <pre> {@code
+   * MessageStream<Long> stream = ...;
+   * MessageStream<WindowOutput<WindowKey, Collection<Long>>> windowed = stream.window(Windows.globalWindow()
+   * .setEarlyTrigger(Triggers.repeat(Triggers.any(Triggers.count(50), Triggers.timeSinceFirstMessage(Duration.seconds(10))))))
+   * }
+   * </pre>
    *
-   * @param sessionKeyFunction  function to calculate session window key
-   * @param <M>  type of the input {@link MessageEnvelope}
-   * @param <WK>  type of the session window key
-   * @return  the {@link Window} function for the session
+   * @param <M> the type of {@link MessageEnvelope}
+   * @return the created {@link WindowFunction} function.
    */
-  public static <M extends MessageEnvelope, WK> Window<M, WK, Integer, WindowOutput<WK, Integer>> intoSessionCounter(Function<M, WK> sessionKeyFunction) {
-    return new SessionWindow<>(sessionKeyFunction, (m, c) -> c + 1);
+
+  public static <M extends MessageEnvelope> WindowFunction<M, Void, WindowKey<Void>, Collection<M>, WindowOutput<WindowKey<Void>, Collection<M>>> globalWindow() {
+    BiFunction<M, Collection<M>, Collection<M>> aggregator = (m, c) -> {
+      c.add(m);
+      return c;
+    };
+    return globalWindow(aggregator);
   }
 
+  /**
+   * Returns a global {@link WindowFunction} that windows its values per-key. The window does not have a default trigger.
+   * The triggering behavior must be specified by setting early
+   * triggers using the {@link org.apache.samza.operators.triggers.Triggers} APIs.
+   *
+   * <p> The below example windows the stream into count based windows. The window triggers every 50 messages or every
+   * 10 minutes.
+   *
+   * <pre> {@code
+   * MessageStream<UserClick> stream = ...;
+   * BiFunction<UserClick, Long, Long> maxAggregator = (m, c)-> Math.max(parseLongField(m), c);
+   * Function<UserClick, String> keyFn = ...;
+   * MessageStream<WindowOutput<WindowKey<String>, Long>> windowed = stream.window(Windows.keyedGlobalWindow(keyFn, maxAggregator)
+   * .setEarlyTrigger(Triggers.repeat(Triggers.any(Triggers.count(50), Triggers.timeSinceFirstMessage(Duration.minutes(10))))))
+   * }
+   * </pre>
+   *
+   * @param keyFn the function to extract the key from a {@link MessageEnvelope}
+   * @param aggregateFunction the function to aggregate window results
+   * @param <M> the type of {@link MessageEnvelope}
+   * @param <K> type of the key in the {@link WindowFunction}
+   * @param <WV> type of the {@link WindowOutput} output value
+   * @return the created {@link WindowFunction} function
+   */
+
+  public static <M extends MessageEnvelope, K, WV> WindowFunction<M, K, WindowKey<K>, WV, WindowOutput<WindowKey<K>, WV>> keyedGlobalWindow(Function<M, K> keyFn, BiFunction<M, WV, WV> aggregateFunction) {
```

Comment:
I think fundamentally I have a confusion here on the proposed global window. Are you proposing to create a window operator that just emits each aggregated window outputs w/o any keys to differentiate the outputs? I.e. if the count is used as aggregate function, the global window will output (null, 3), (null, 5), (null, 9)...? How do you assume the downstream consumer to consume this sequence of window output? In UMP's example of "global" window, at least, the count values are assume to be labeled by timestamp, i.e. (0'0'', 3), (0'1", 5), (0'2", 9) ... although not keyed by a certain field in the incoming message. I can think of use cases that does not need the key function, but really have difficulties in finding a use case that does not need the pane ids in the output. Can we discuss more on this one?